programs/cryptid/src/instructions/propose_transaction.rs

```diff
@@ -86,66 +85,52 @@ pub fn propose_transaction<'a, 'b, 'c, 'info>(
-        })
-        .collect::<Vec<(&AccountInfo, Pubkey)>>();
-
-    // Assume at this point that anchor has verified the cryptid account and did account (but not the controller chain)
-    // We now need to verify that the signer (at the moment, only one is supported) is a valid signer for the cryptid account
-    verify_keys(
-        &ctx.accounts.did,
-        Some(did_account_bump),
-        ctx.accounts.authority.to_account_info().key,
-        controlling_did_accounts,
-    )?;
 
-    // For seed verification
-    CryptidAccount::try_from(
+    get_cryptid_account_checked(
+        &all_accounts,
+        &controller_chain,
         &ctx.accounts.cryptid_account,
-        &ctx.accounts.did_program.key(),
-        &ctx.accounts.did.key(),
+        &ctx.accounts.did,
+        &ctx.accounts.did_program,
+        &ctx.accounts.authority,
+        did_account_bump,
         cryptid_account_index,
         cryptid_account_bump,
     )?;
-
-    ctx.accounts.transaction_account.did = *ctx.accounts.did.key;
-    ctx.accounts.transaction_account.instructions = instructions;
-    ctx.accounts.transaction_account.cryptid_account = *ctx.accounts.cryptid_account.key;
-    ctx.accounts.transaction_account.approved_middleware = None;
-    // Extending transactions is not yet supported, so transactions are initialised in Ready state
-    ctx.accounts.transaction_account.state = TransactionState::Ready;
-
     // Accounts stored into the transaction account are referenced by
     // the abbreviated instruction data by index
     // The same accounts must be passed, in the correct order, to the ExecuteTransaction instruction
     // Note - the order is retained between Propose and Execute, but some accounts are omitted during Propose
     //
     // Execute Transaction Accounts:
     // 0 - cryptid account
-    // 1 - did*
-    // 2 - did program*
-    // 3 - signer*
+    // 1 - did
+    // 2 - did program
+    // 3 - signer
     // ... remaining accounts
     //
-    // * These accounts are omitted from the Propose Transaction Accounts
     // Account indexes must reflect this, so the first entry
     // in the remaining accounts is referred to in the abbreviated instruction data as index 4,
     // despite being index 0 in the remaining accounts.
     // TODO validate that the account indices are all valid, given the above i.e. that no index exceeds remaining_accounts.length + 4
-    ctx.accounts.transaction_account.accounts =
-        ctx.remaining_accounts.iter().map(|a| *a.key).collect();
+    ctx.accounts.transaction_account.accounts = all_accounts.iter().map(|a| *a.key).collect();
+
+    ctx.accounts.transaction_account.did = *ctx.accounts.did.key;
+    ctx.accounts.transaction_account.instructions = instructions;
+    ctx.accounts.transaction_account.cryptid_account = *ctx.accounts.cryptid_account.key;
+    ctx.accounts.transaction_account.approved_middleware = None;
+
+    // we cannot initiate a transaction in executed state.
+    require_neq!(
```

Comment:
Ok I understand this is not part of this PR, but why do we need the state `ctx.accounts.transaction_account.state = TransactionState::Executed;` at all (and therefore check for it here)? The transaction_account get's closed right after successful execution, so i my mind the state (and the check is not needed?)

Reply:
as agreed, we keep it in for now.